viewer-prototype/src/browser/trace-viewer/components/utils/filter-tree-component.tsx

```diff
@@ -0,0 +1,230 @@
+    }
+
+    getNode: any = (node: TreeNode, id: number) => {
+        if (node.id === id) {
+            return node;
+        } else if (node.children !== undefined && node.children.length > 0) {
+            let result = null;
+            for (let i = 0; result == null && i < node.children.length; i++) {
+                result = this.getNode(node.children[i], id);
+            }
+            return result;
+        }
+        return null;
+    }
+
+    onToggle = (node: TreeNode) => {
+        const nodes = this.state.treeNodes;
+        const toggledNode = this.getNode(nodes[0], node.id);
+        toggledNode.expanded = !node.expanded;
+        this.setState({treeNodes: nodes})
+    }
+
+    getAllChildrenIds = (node: TreeNode, ids: number[]) => {
+        if (node.type === 'leaf') {
+            return ids.push(node.id);
+        } else if (node.children.length) {
+            ids.push(node.id);
+            node.children.forEach((child: TreeNode) => {
+                this.getAllChildrenIds(child, ids);
+            })
+        }
+        return ids;
+    }
+
+    areAllSiblingsChecked = (node: TreeNode) => {
+        const currentCheckedIds = this.state.checkedIDs.concat([node.id]);
+        const parentNode = this.getNode(this.state.treeNodes[0], node.parentId);
+        return parentNode.children.every((child: TreeNode) => {
+            return currentCheckedIds.includes(child.id);
+        });
+    }
+
+    handleCheck = (id: number) => {
+        let checkedIds: number[] = [];
+        const checkedNode = this.getNode(this.state.treeNodes[0], id);
+        if (checkedNode.type === 'node' && checkedNode.children.length) {
+            checkedIds = checkedIds.concat(this.getAllChildrenIds(checkedNode, []));
+        } else {
+            checkedIds.push(id);
+            if (this.areAllSiblingsChecked(checkedNode)) {
+                checkedIds.push(checkedNode.parentId);
+            }
+        }
+        this.props.onChecked(checkedIds);
+    }
+
+    render() {
+        if (!this.props.nodes || !this.props.checkedSeries) return null;
+        const rootNodes = this.getRootNodes();
+        if (rootNodes !== undefined && rootNodes.length > 0) {
+            return (
+                <div>
+                    {rootNodes.map((node, i) => {
+                        return (
+                            <TreeNodeComponent 
+                                node={node}
+                                key={'node-'+node.id}
+                                getChildNodes={this.getChildNodes}
+                                level={0}
+                                onToggle={this.onToggle}
+                                checkedSeries={this.props.checkedSeries}
+                                onChecked={this.handleCheck}
+                            />
+                        )
+                        
+                    })}
+                </div>
+            )
+        }
+        return (<div></div>)
```

Comment:
Here, you could return <EmptyTree> and create a component for it so we can edit it later when we want something specific